07-memory/user_space/main.c

```diff
@@ -0,0 +1,109 @@
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+#define NVALUE 64
+#define NSTAT 10
+
+static void malloc_inv(int ind, int cal, int p1)
+{
+	clock_t tm, tf;
+	void *ptr;
+	double btm = 0;
+	double btf = 0;
+	double min_tm = 0;
+	double max_tm = 0;
+	double min_tf = 0;
+	double max_tf = 0;
+
+	double size = pow(2, ind);
+	if (p1) {
+		size++;
+	}
+
+	for (int i = 0; i <  NSTAT; i++) {
+
+		if (cal == 2) {
+			tm = clock();
+			ptr = alloca(size * sizeof(void));
+			tm = clock() - tm;
+		} else if (cal == 1) {
+			tm = clock();
+			ptr = calloc(size, sizeof(void));
+			tm = clock() - tm;
+		} else {
+			tm = clock();
+			ptr = malloc(size * sizeof(void));
+			tm = clock() - tm;
+		}
+
+		if (ptr == NULL) {
+			printf("%2d %22.0f: Memory not allocated.\n", ind, size);
+			return;
+		}
+
+		btm += ((double)tm);
+		if ((double)tm < min_tm)
+			min_tm = (double)tm;
+		if ((double)tm > max_tm)
+			max_tm = (double)tm;
+
+		if (cal != 2) {
+			tf = clock();
+			free(ptr);
+			tf = clock() - tf;
+			btf += ((double)tf);
+
+			if ((double)tf < min_tf)
+				min_tf = (double)tf;
+			if ((double)tf > max_tf)
+				max_tf = (double)tf;
+		}
+
+	}
+	printf("%2d %22.0f %10.0f %10.0f %10.0f %10.0f %10.0f %10.0f\n",
+	ind, size, min_tm, btm/NSTAT, max_tm, min_tf, btf/NSTAT, max_tf);
+}
+
+int main(int argc, char *argv[])
+{
+	int i;
+
+	printf("************************\n");
+	printf("* malloc function, 2^n *\n");
+	printf("************************\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 0, 0);
+
+	printf("************************\n");
+	printf("* calloc function, 2^n *\n");
+	printf("************************\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 1, 0);
+
+// Segmentation fault:
+/*	printf("alloca function, 2^n\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 2, 0);*/
+
+	printf("***************************\n");
+	printf("* malloc function, 2^n+1  *\n");
+	printf("***************************\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 0, 1);
+
+	printf("****************************\n");
+	printf("* calloc function, 2^n+1   *\n");
+	printf("****************************\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 1, 1);
+
+// Segmentation fault:
+/*	printf("alloca function, 2^n+1\n");
+	for (i = 0; i < NVALUE; i++)
+		malloc_inv(i, 2, 1);*/
+
```

Comment:
For prevent segmentation fault it is possible to limit NVALUE for alloca. e.g. to 18.